### ISSUE TITLE
Dumping the core on hitting a JASSERT?

### ISSUE THREAD

**Issue by rohgarg** [dmtcp/dmtcp#112]
I have noticed that JASSERT doesn't generate a coredump, which seems to suggest that it calls exit().

Can we add a debug env var and/or a debug flag which triggers an abort or a SEGFAULT on a JASSERT? This would make debugging easier in some cases (most notably when we have distributed processes).

**Comment by gc00** [comment]
@rohgarg: Good point.  It's not always easy to run under GDB with a breakpoint at `_exit()`.  This is often more convenient.

Rohan, if you'd like to generate a pull request, here's an easy recipe for the implementation.  Create a `JASSERT(false);` somewhere in `dmtcp_launch`.  Then do:
```gdb --args dmtcp_launch test/dmtcp1
(gdb) break _exit
(gdb) run
(gdb) bt```
By looking at the stack, it should be obvious how jalib finally calls `exit()`.
Then insert code that checks a DMTCP environment variable (DMTCP_ABORT_ON_FAILED_ASSERT ??) to see if it's set, and if so, call `abort()`.  Then propose a pull request.

Also, I've updated doc/debugging-dmtcp.txt in the branch `gdb-attach-on-restart`.  Hopefully, that will get pushed in soon.  Could you add an appropriate documentation in that document as part of the pull request?  Thanks.

**Comment by gc00** [comment]
@karya, @rohgarg:  Another alternative is to have JASSERT always dump core.  Most users have coredumpsize set to 0 by default, and only raise it if they want to see a core dump.  So should DMTCP always dump core on JASSERT (similarly to the standard 'assert' used in C/C++)?

**Comment by rohgarg** [comment]
@gc00: Kapil and I were discussing this earlier, and he observed that there's a possibility of the core-dump filling up the disk space when running large computations.

We decided that using `exit()` by default makes it safe for most users, and for the more advanced users, we can provide the debug environment variable.

**Comment by gc00** [comment]
That makes a lot of sense.  JASSERT could be used with large HPC programs.

Anyway, I think with the recipe above, it should be easy to generate a new pull request for this.

**Comment by rohgarg** [comment]
Closed via ee4a864.
